### PR TITLE
Re-enable optimizers

### LINKED ISSUE #90165: Optimizer operations as a graph
### 🚀 The feature, motivation and pitch

Hello everyone, I hope you're having a glorious day. I'm coming with a question/proposition/bug, I'm not sure which category this should be in, so I'm putting it as feature request.

I just checked latest PT 2.0 code version and tried to use some kind of simple LeNet topology with new torch.compile function.

image

(I'm running this on CPU BTW)

I decided to do tracing of above code to see how well it creates the graphs from this function. Results looks like this:
image

As you can see, FWD and BWD became graph optimized by Inductor backend. This is what we would like to achieve usually on high-end accelerator, right? But, on high-end accelerator we wouldn't like to have basically any operations outside of the graphs, and as you can see there, in the beginning there is grad zeroing which doesn't seem to be a part of any graph (is ran eagerly), same for grad accumulation and what is the worst here - optimizer itself.

There are topologies with quite heavy optimizers and this would cause accelerator to be very poorly utilized here, hence my questions:

1. Are you planning to support compilation of optimizer part? I'd like to be able to compile it in the backend (e.g. Inductor) same way is done with FWD/BWD. Could be as third graph, for example.
2. Generalizing on above - are you planning to support whole-function graph optimization? Ideally, I'd like to see WHOLE function as a single merged graph at some point in the backend to get maximum optimization opportunities (with some exceptions, but this is what partitioners are for, right? Sometimes it would be best to work on FWD, BWD, OPT graphs, and sometimes on FWD+BWD, OPT graphs, and so on...). I'm not saying it should be default behaviour - I'd just like to see it being an option.

If any of those can be done already now and it's just me not knowing how to use compilation properly, please tell me!

Cheers,
--mb

### Alternatives

_No response_

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @janeyx99 @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### PR DESCRIPTION
**Pull request by mlazos**
Fixes 
https://github.com/pytorch/pytorch/issues/90165
https://github.com/pytorch/pytorch/issues/93758

Re-enables optimizer capture + compilation now that the dynamo slowdowns have been fixed

and it has speedups, numbers to come soon 

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by anijain2305 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Review by yanboliang — APPROVED** [review]

**Comment by desertfire** [comment]
Heads up on https://github.com/pytorch/torchdynamo/issues/1985. Turning on optimizer causes some random accuracy failures. Please work with Animesh to fix that first.

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -764,8 +764,22 @@ def patch():
...
+                setattr(
+                    opt_mod,
+                    multi_tensor_fn_name,
+                    disable(getattr(opt_mod, multi_tensor_fn_name)),
+                )
+
+        excluded_opts = {torch.optim.SparseAdam, torch.optim.RAdam, torch.optim.LBFGS}
         for opt in optimizers:
+            if opt in excluded_opts:
+                disable(opt.step)
```
[the remark this answers is not in the document]
Comment (mlazos) [reply torch/_dynamo/eval_frame.py:781]: Yes, good catch!

**Comment by mlazos** [comment]
> Heads up on https://github.com/pytorch/torchdynamo/issues/1985. Turning on optimizer causes some random accuracy failures. Please work with Animesh to fix that first.

Sounds good. I will take a look. Oh but we're literally running in eager mode at the moment
